### PR TITLE
Pakurapo/to pr

### PR DESCRIPTION
**Pull request by kurapov-peter**

### CODE REVIEW

File: QueryEngine/NativeCodegen.cpp [hunk 1]
```diff
@@ -1492,53 +1621,54 @@ void set_row_func_argnames(llvm::Function* row_func,
...
                                     llvm::Module* module,
                                     llvm::LLVMContext& context) {
   std::vector<llvm::Type*> row_process_arg_types;
+  unsigned int AS = (co.device_type == ExecutorDeviceType::L0) ? 4 : 0;
 
   if (agg_col_count) {
     // output (aggregate) arguments
     for (size_t i = 0; i < agg_col_count; ++i) {
-      row_process_arg_types.push_back(llvm::Type::getInt64PtrTy(context));
+      row_process_arg_types.push_back(llvm::Type::getInt64PtrTy(context, AS));
```
Comment (fexolm) [inline QueryEngine/NativeCodegen.cpp:1633]: What is AS?

File: QueryEngine/NativeCodegen.cpp [hunk 2]
```diff
@@ -1492,53 +1621,54 @@ void set_row_func_argnames(llvm::Function* row_func,
 
 llvm::Function* create_row_function(const size_t in_col_count,
                                     const size_t agg_col_count,
-                                    const bool hoist_literals,
+                                    const CompilationOptions& co,
```
Comment (fexolm) [inline QueryEngine/NativeCodegen.cpp:1624]: It doesn't look like omnisci use that short naming convention. Maybe rename it to compilation_options?

File: QueryEngine/NativeCodegen.cpp [hunk 3]
```diff
@@ -2602,6 +2739,9 @@ Executor::compileWorkUnit(const std::vector<InputTableInfo>& query_infos,
       CodeGenerator::link_udf_module(
           rt_udf_gpu_module, *rt_module_copy, cgen_state_.get());
     }
+  } else {
```
Comment (fexolm) [inline QueryEngine/NativeCodegen.cpp:2742]: mb switch with default check false?

File: QueryEngine/ScalarCodeGenerator.cpp [hunk 4]
```diff
@@ -74,6 +74,7 @@ ScalarCodeGenerator::CompiledExpression ScalarCodeGenerator::compile(
     const Analyzer::Expr* expr,
     const bool fetch_columns,
     const CompilationOptions& co) {
+  unsigned int AS = (co.device_type == ExecutorDeviceType::L0) ? 1 : 0;
```
Comment (fexolm) [inline QueryEngine/ScalarCodeGenerator.cpp:77]: addr_space is a bit more clear, I think

File: scripts/common-functions.sh [hunk 5]
```diff
@@ -100,7 +100,7 @@ function install_arrow() {
     -DARROW_PARQUET=ON \
     -DARROW_FILESYSTEM=ON \
     -DARROW_S3=ON \
-    -DARROW_CUDA=ON \
+    -DARROW_CUDA=OFF \
```
Comment (fexolm) [inline scripts/common-functions.sh:103]: It looks like this change also affects omnmisci cuda execution. We need to bypass it somehow

File: scripts/mapd-deps-ubuntu.sh [hunk 6]
```diff
@@ -43,11 +43,11 @@ else
...
 fi
 
-sudo mkdir -p $PREFIX
-sudo chown -R $(id -u) $PREFIX
+ mkdir -p $PREFIX
+ chown -R $(id -u) $PREFIX
 
-DEBIAN_FRONTEND=noninteractive sudo apt update
-DEBIAN_FRONTEND=noninteractive sudo apt install -y \
+DEBIAN_FRONTEND=noninteractive  apt update
```
Comment (fexolm) [inline scripts/mapd-deps-ubuntu.sh:49]: this also affects omnisci

File: QueryEngine/NativeCodegen.cpp (hunk 2 above)
Comment (ienkovich) [reply QueryEngine/NativeCodegen.cpp:1624]: Actually, it's heavily used in the sources. `co` for compilation options and `eo` for execution options.

File: QueryEngine/NativeCodegen.cpp (hunk 1 above)
Comment (kurapov-peter) [reply QueryEngine/NativeCodegen.cpp:1633]: That's more or less standard, I don't mind renaming though.

File: scripts/common-functions.sh (hunk 5 above)
Comment (kurapov-peter) [reply scripts/common-functions.sh:103]: Yup, that's from the "todo" commit. Any ideas on resolving? ;)

File: QueryEngine/CMakeLists.txt [hunk 7]
```diff
@@ -274,6 +280,12 @@ set(QUERY_ENGINE_LIBS
 
 list(APPEND QUERY_ENGINE_LIBS ${clang_libs})
 list(APPEND QUERY_ENGINE_LIBS ${llvm_libs} ${ZLIB_LIBRARIES})
+list(APPEND QUERY_ENGINE_LIBS LLVMSPIRVLib)
+
+if(ENABLE_L0)
+  list(APPEND QUERY_ENGINE_LIBS L0Mgr) # ??
+#  list(APPEND QUERY_ENGINE_LIBS ${LevelZero_COMPONENTS})
```
Comment (kurapov-peter) [inline QueryEngine/CMakeLists.txt:287]: cleanup

File: QueryEngine/CodeGenerator.h [hunk 8]
```diff
@@ -608,7 +618,13 @@ class ScalarCodeGenerator : public CodeGenerator {
   std::vector<void*> generateNativeCode(const CompiledExpression& compiled_expression,
                                         const CompilationOptions& co);
 
+  // Specific method for L0 code generation - L0 kernel is called with a handle.
```
Comment (kurapov-peter) [inline QueryEngine/CodeGenerator.h:621]: unnecessary

File: QueryEngine/L0Kernel.cpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+#include <level_zero/ze_api.h>
+
+L0BinResult spv_to_bin(const std::string& spv,
+                       const std::string& name,
+                       const unsigned block_size,
+                       const l0::L0Manager* mgr) {
+  CHECK(!spv.empty());
+  CHECK(mgr);
+
+  // void* bin{nullptr};
```
Comment (kurapov-peter) [inline QueryEngine/L0Kernel.cpp:18]: cleanup

File: QueryEngine/NativeCodegen.cpp [hunk 10]
```diff
@@ -1163,6 +1169,129 @@ std::shared_ptr<GpuCompilationContext> CodeGenerator::generateNativeGPUCode(
...
+    const l0::L0Manager* l0_mgr) {
+  auto module = func->getParent();
+
+  auto pass_manager_builder = llvm::PassManagerBuilder();
+  llvm::legacy::PassManager PM;
+  pass_manager_builder.populateModulePassManager(PM);
+  optimize_ir(func, module, PM, live_funcs, co);
+
+  // PM.add(new LegacyOCLAddressLegalizerPass());
+  // PM.run(*module);
```
Comment (kurapov-peter) [inline QueryEngine/NativeCodegen.cpp:1187]: cleanup

File: QueryEngine/NativeCodegen.cpp [hunk 11]
```diff
@@ -1163,6 +1169,129 @@ std::shared_ptr<GpuCompilationContext> CodeGenerator::generateNativeGPUCode(
...
+      rt_funcs.push_back(&Fn);
+    }
+  }
+
+  for (auto& pFn : rt_funcs) {
+    // pFn->removeFromParent();
+    pFn->eraseFromParent();
+  }
+
+  // for (auto& pFn : rt_funcs) {
```
Comment (kurapov-peter) [inline QueryEngine/NativeCodegen.cpp:1227]: need at least comment this

File: Tests/SpirvExecuteTest.cpp [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+
+  ASSERT_EQ(b.data[0], 16);
+  ASSERT_EQ(b.data[1], 6);
+  ASSERT_EQ(b.data[2], 6);
+
+  L0_SAFE_CALL(zeMemFree(device->ctx(), dA));
+  L0_SAFE_CALL(zeMemFree(device->ctx(), dB));
+}
+
+TEST_F(SPIRVExecuteTest, TranslateSimpleWithL0Manager) {
```
Comment (kurapov-peter) [inline Tests/SpirvExecuteTest.cpp:338]: looks like I've duplicated the test

**Pull request merged**